kotlinx-coroutines-core/common/src/internal/SegmentQueueSynchronizer.kt

```diff
@@ -0,0 +1,606 @@
+ * after that. Following this pattern, it is possible in a concurrent environment that [resume]
+ * is executed before [suspend] (similarly to the race between `park` and `unpark` for threads),
+ * so that [resume] comes to a cell in the empty state. This race can be solved with two [strategies][ResumeMode]:
+ * [asynchronous][ASYNC] and [synchronous][SYNC]. In the [asynchronous][ASYNC] mode, [resume] puts
+ * the value into the cell if it is still empty and finishes immediately; this way, the following [suspend] comes
+ * to this cell and simply grabs the element without suspension. At the same time, in the [synchronous][SYNC] mode,
+ * [resume] waits in a bounded spin-loop cycle until the put element is taken by a concurrent [suspend],
+ * marking the cell as broken if it is not taken after the spin-loop is finished. In this case, both the current
+ * [resume] and the [suspend] that comes to this broken cell fail and the corresponding operations on the data
+ * structure are typically restarted from the beginning.
+ *
+ * Since [suspend] can store [CancellableContinuation]-s, it is possible for [resume] to fail if the
+ * continuation is already cancelled. In this case, most of the algorithms retry the whole operation.
+ * However, if there are many consecutive cancelled waiters, it seems more efficient to skip them
+ * somehow. Thus, there are different [cancellation modes][cancellationMode] that can be used.
+ * In the simple mode ([SIMPLE]), which is used by default, [resume] simply fails on cancelled waiters as
+ * described above, while in smart modes ([SMART_SYNC] and [SMART_ASYNC]) [resume] skips cells in the
+ * [cancelled][CANCELLED] state. However, if cancellation happens concurrently with [resume], it can be illegal
+ * to simply skip the cell and resume the next waiter, e.g., if this cancelled waiter is the last one.
+ * Thus, it is possible for [SegmentQueueSynchronizer] to refuse this [resume]. In order to support this logic,
+ * users should implement  [onCancellation] function, which returns `true` if the cell can be
+ * moved to the [cancelled][CANCELLED] state, or `false` if the [resume] that comes to this cell
+ * should be refused. In the last case, [tryReturnRefusedValue] is invoked, so that the value
+ * can be put back to the data structure. However, it is possible for [tryReturnRefusedValue] to
+ * fail, and [returnValue] is called in this case. Typically, this [returnValue] function coincide
+ * with the one that resumes waiters (e.g., [release][Semaphore.release] in [Semaphore]). The difference
+ * between [SMART_SYNC] and [SMART_ASYNC] modes is that in the first one the [resume] that comes to the
+ * cell with cancelled waiter waits in a spin-loop until the cell is marked either with [CANCELLED]
+ * or [REFUSE] mark, while in [SMART_ASYNC] mode it replaces the cancelled waiter with the value of
+ * the resumption and finishes, so that the cancellation handler completes this [resume] eventually.
```

Comment:
There is little rationale for why the two modes are needed. The next statement does mention that [SMART_ASYNC] may be undesirable due to problems with linearization, but the design of these modes does not follow at all. Additional clarification is needed, like mentioning that [SMART_SYNC] does this so that by the end of the operation the element is always done being passed to the data structure.